src/Hamiltonians/abstract.jl

```diff
@@ -139,56 +139,79 @@ julia> rayleigh_quotient(mom, v) # momentum expectation value for state vector `
 momentum
 
 """
-    sm, basis = build_sparse_matrix_from_LO(ham::AbstractHamiltonian, add; nnzs = 0)
+    sm, basis = build_sparse_matrix_from_LO(ham, add; cutoff, nnzs)
```

Comment:
```suggestion
    sm, basis = build_sparse_matrix_from_LO(ham, add; cutoff, filter, nnzs)
```
I'd like to suggest to make it possible to pass a user-defined function `filter(address) -> ::Bool` that would allow users to define their own filters to selectively truncate the matrix. If `filter(address)` returns `true` the address is accepted and otherwise skipped.

The default could be 
```
filter = addr -> diagonal_element(ham, addr) < cutoff
```
and setting the `filter` keyword would have precedence over setting `cutoff`.

This would allow the easy implementation of various flavours of [truncated CI](https://en.wikipedia.org/wiki/Configuration_interaction), which is very popular in quantum chemistry, e.g. SDCI where configurations are restricted to single and double excitations from the Hartree-Fock Slater determinant (our `starting_address`).

Question: Would the suggested change have an impact on runtime performance and, e.g., lead to dynamic dispatch? I hope it would not, but I'm sometimes confused about what the compiler does with keyword arguments.

Another comment is that the additional keyword arguments should be documented for `BasisSetRep`, which is the main user entry point, and possibly for the methods we define for `sparse` and `Matrix`.